Extend Contribution Guide

lukaszstolarczuk opened the pull request:
We should consider this as the first version of this document. It should be updated with more content as we go.

igchor commented on .github/ISSUE_TEMPLATE/bug_report.md:
```diff
@@ -0,0 +1,50 @@
+
+## Environment Information
+
+- UMF version(s):                                             <!-- fill this out -->
```
Perhaps a 'git tag' for now since we don't really have any versions released yet?

igchor commented on CONTRIBUTING.md:
```diff
@@ -1,8 +1,74 @@
+**NOTE**: We use specific clang-format - version exactly **15.0** is required.
+
+<!--
+### Naming Convention
```
That's a good question ;d Originally we had the convention as in UR: camelCase for function names and snake_case for type names. However, I think now we also have some files where functions are named using snake_case - it would be nice to unify that.

bratpiorka replied:
Please remove this section for the moment - I added a JIRA UMFW-248 for changing code to use single naming convention in all files + adding a section about it here to this doc

igchor commented on CONTRIBUTING.md:
```diff
@@ -55,3 +121,28 @@ Use your real name (sorry, no pseudonyms or anonymous contributions.)
+
+
+<!--
+this is copy-pasted from pmemstream, it was used for OSPDT updates
```
I think this is a good practice so I would leave it.

PatKamin commented on CONTRIBUTING.md:
```diff
@@ -1,8 +1,74 @@
+    - [Adding new dependency](#adding-new-dependency)
+    - [Extending Public API](#extending-public-api)
+
+Below you'll find instructions on how to contribute to UMF, either with code changes
```
Using either with more than two alternatives makes me unease. Maybe rephrase it a bit:
```
Below you'll find instructions on how to contribute to UMF. This can be done in three ways:
- open a PR with code changes
- file an issue with a problem to solve
- file an issue with an enhancement proposal.
```

PatKamin commented on CONTRIBUTING.md:
```diff
@@ -1,8 +1,74 @@
+
+.clang-format
+UMF_FORMAT_CODE_STYLE
+make clang-format-check & apply
```
Use portable CMake commands.

PatKamin commented on CONTRIBUTING.md:
```diff
@@ -1,8 +1,74 @@
+
+### Building and testing
+
+described in Readme...
```
Remember to add a link to README (or even the exact section of it).

PatKamin commented on CONTRIBUTING.md:
```diff
@@ -1,8 +1,74 @@
+make clang-format-check & apply
+DEVELOPER_MODE
+
+**NOTE**: We use specific clang-format - version exactly **15.0** is required.
```
(...) clang-format version - exactly (...)

lukaszstolarczuk commented:
I've rebased, made a little update in Readme file (added a link to CONTRIBUTING.md) and updated a little content of code-style and test-and-build sections.

PatKamin commented on CONTRIBUTING.md:
```diff
@@ -1,12 +1,100 @@
+- [Submitting Pull Requests](#submitting-pull-requests)
+    - [Building and testing](#building-and-testing)
+    - [Code style](#code-style)
+    <!-- - [Naming convention](#naming-convention) -->
```
Either remove this or move to the top of this list.

lukaszstolarczuk replied:
why? it doesn't break the list and when we add this section it's going to be right about here 😉

PatKamin commented on CONTRIBUTING.md:
```diff
@@ -1,12 +1,100 @@
+please take a look at `.clang-format` file in the root directory of this repository.
+
+To enable code style checks and re-formatting, CMake option `UMF_FORMAT_CODE_STYLE` has to
+be switched on. You'll then have two additional commands available.
```
commands -> CMake targets ?

lukaszstolarczuk replied:
Done.

PatKamin commented on CONTRIBUTING.md:
```diff
@@ -1,12 +1,100 @@
 ### Sign your work
 
-Please use the sign-off line at the end of the patch. Your signature certifies that you wrote the patch or otherwise have the right to pass it on as an open-source patch. The rules are pretty simple: if you can certify
+Please use the sign-off line at the end of the patch. Your signature certifies that you wrote the patch
```
I thought that signing-off wasn't meant to be a requirement in this project, @bratpiorka ?

lukaszstolarczuk replied:
It's still not a requirement - it's merely an ask 😉 Also, for people in our organization this may not be needed, but perhaps for outside contributors, it's a good thing to ask...?

bratpiorka replied:
I would remove it now. I am not sure if even my approvals are signed. This is also not tested in CI. IMO this whole file contains not questions, but requirements that would fail CI if not met.

bratpiorka replied:
Again - could we test this? Even if not now, then in the future? If we leave it, the whole UMF team should respect it and we should not merge any non-signed-off code. But if it is a 'nice to have', I would not mention it to avoid unnecessary misunderstandings and discussions.

lukaszstolarczuk replied:
heh, we had a clause like this in all(?) our previous repos, and I cannot remember a single discussion about it 😉 
We can test anything, including this (but I hope there's a tool for that 😉 ).

// examples: [PMDK](https://github.com/pmem/pmdk/blob/master/CONTRIBUTING.md) / [pmemkv](https://github.com/pmem/pmemkv/blob/master/CONTRIBUTING.md)

That said, we can use similar wording to one in PMDK, to be more flexible about it:

>In case of any doubt, the gatekeeper may ask you to certify the above in writing, i.e. via email or by including a Signed-off-by: line at the bottom of your commit comments.

bratpiorka replied:
"Anecdotal evidence" :D I'm ok with "In case of any doubt" form. If we want to make it stronger we have to follow it + add tests.

lukaszstolarczuk replied:
Rewritten.

bratpiorka commented on CONTRIBUTING.md:
```diff
@@ -1,12 +1,100 @@
+include only relevant ones (no "fixes after review" or similar).
+
+Before your pull request is merged it *must* pass all jobs in the GitHub
+Actions workflows and *must* be reviewed by no less than two reviewers.
```
Please add a section about the review. Add info that "codeowners" would get a notification when PR is created, but one could always add specific people if needed which should speed up the review process. Also add info that PR is merged when CI passes, all discussions are resolved and PR got at least two approves.

bratpiorka commented on CONTRIBUTING.md:
````diff
@@ -1,12 +1,100 @@
+$ cmake --build build --target clang-format-check
+```
+
+If any issues were found, apply proper formatting via command:
````
Add that the issues in formatting will fail CI and block the merge

bratpiorka commented on CONTRIBUTING.md:
```diff
@@ -1,12 +1,100 @@
 
-<PROJECT NAME> is licensed under the terms in [LICENSE]<link to license file in repo>. By contributing to the project, you agree to the license and copyright terms therein and release your contribution under these terms.
+Unified Memory Framework is licensed under the terms in [LICENSE](./LICENSE.TXT). By contributing to the project,
+you agree to the license and copyright terms therein and release your contribution under these terms.
```
Add that all new files should start from a license.

bratpiorka commented on CONTRIBUTING.md:
```diff
@@ -1,12 +1,134 @@
+
+### When my PR is merged?
+
+To get your Pull Request (PR) merged you have to fulfill several requirements.
```
I'm not an expert but I think this could sound better:

To get your Pull Request (PR) to be merged, you must meet several requirements - the basic are:
- The project compiles correctly,
- ....

lukaszstolarczuk replied:
This way is less fun 😄😉 Rewritten.

PatKamin commented on CONTRIBUTING.md:
```diff
@@ -1,13 +1,129 @@
 
-### Sign your work
+**Note:**
+>Each new file added to the repository has to contain the appropriate license header. To see how such header
```
such (a?) header **looks** like

lukaszstolarczuk replied:
Done.

PatKamin commented on CONTRIBUTING.md:
```diff
@@ -1,13 +1,129 @@
+While the most of these requirements are verified via automated scripts run in
+Continuous Integration (CI) they are also verified with human touch - **the review**!
+
+Code review has to be done by at least two UMF maintainers. They "maintainers team"
```
They -> The

lukaszstolarczuk replied:
Done.
